### PR TITLE
Select statements and Session

### PR DESCRIPTION
**Pull request by guynir42**
his PR introduces scoped sessions with user access verification.

It also introduces new methods on the Base model, to allow querying using the new "select" style compatible with SQLA 2.0

It also fixes the behavior of `get_if_accessible_by` to respond exactly the same if the row requested does not exist or if it is not accessible. 

This is mostly backward compatible, but a separate PR to skyportal fixes the few issues in the handlers and the tests that are broken by this change. 

Most importantly, it allows us to move each handler separately into the new SQLA 2.0 style without having to change everything at once.

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -20,14 +21,155 @@
...
+    Create an instance of Session when you
+    want to apply a verification method on all added
+    or modified or deleted rows before commiting them
+    to the database.
+
+    This class overrides the commit() function
+    by adding a verify() function before it.
+
+    Use this in a context manager:
+    with VerifiedSession(user_object) as session:
```
Comment (stefanv) [inline app/models.py:45]: Are you still planning to rename to VerifiedSession as indicated here?

File: app/models.py [hunk 2]
```diff
@@ -20,14 +21,155 @@
...
+        ...
+        session.commit()
+
+    This will make sure the changes to the DB
+    are verified, and will close the connection
+    when the context goes out.
```
Comment (stefanv) [inline app/models.py:51]:
```suggestion
    when exiting the context.
```

File: app/models.py [hunk 3]
```diff
@@ -20,14 +21,155 @@
...
+            ).first()
+
+        Parameters
+        ----------
+        user_or_token: baselayer.app.models.User object
+            or baselayer.app.models.Token object.
+            The object representing the current user.
+
+        """
+        self.user_or_token = user_or_token
```
Comment (stefanv) [inline app/models.py:74]: Remember to also call `super().__init__()`

File: app/models.py [hunk 4]
```diff
@@ -20,14 +21,155 @@
...
+        self.flush()
+        bulk_verify("create", new_rows, self.user_or_token)
+
+    def commit(self):
+        print("verifying rows before committing... ")
+        self.verify()
+        super().commit()
+
+
+VerifiedSession = scoped_session(
```
Comment (stefanv) [inline app/models.py:121]: It's confusing to have Session and VerifiedSession being entirely different things. Maybe the class should be called `_VerifiedSession` to make clear that `VerifiedSession` is the one to use?

File: app/models.py [hunk 5]
```diff
@@ -305,6 +471,30 @@ def query_accessible_rows(self, cls, user_or_token, columns=None):
             return DBSession().query(*columns).select_from(cls)
         return DBSession().query(cls)
 
+    def select_accessible_rows(self, cls, user_or_token, columns=None):
+        """Construct a Select object that, when executed, returns the rows of a
+        specified table that are accessible to a specified user or token.
+
+        Parameters
+        ----------
+        cls: `baselayer.app.models.DeclarativeMeta`
```
Comment (stefanv) [inline app/models.py:480]:
```suggestion
        cls : `baselayer.app.models.DeclarativeMeta`
```

File: app/models.py [hunk 6]
```diff
@@ -305,6 +471,30 @@ def query_accessible_rows(self, cls, user_or_token, columns=None):
...
+            The mapped class of the target table.
+        user_or_token: `baselayer.app.models.User` or `baselayer.app.models.Token`
```
Comment (stefanv) [inline app/models.py:482]:
```suggestion
        user_or_token : `baselayer.app.models.User` or `baselayer.app.models.Token`
```

File: app/models.py [hunk 7]
```diff
@@ -305,6 +471,30 @@ def query_accessible_rows(self, cls, user_or_token, columns=None):
...
+            The User or Token to check.
+        columns: list of sqlalchemy.Column, optional, default None
```
Comment (stefanv) [inline app/models.py:484]:
```suggestion
        columns : list of sqlalchemy.Column, optional, default None
```

File: app/models.py [hunk 8]
```diff
@@ -389,6 +581,51 @@ def query_accessible_rows(self, cls, user_or_token, columns=None):
         query = query.filter(cls.id == user_id)
         return query
 
+    def select_accessible_rows(self, cls, user_or_token, columns=None):
+        """Construct a Select object that, when executed, returns the rows of a
+        specified table that are accessible to a specified user or token.
+
+        Parameters
+        ----------
+        cls: `baselayer.app.models.DeclarativeMeta`
```
Comment (stefanv) [inline app/models.py:590]:
```suggestion
        cls : `baselayer.app.models.DeclarativeMeta`
```

File: app/models.py [hunk 9]
```diff
@@ -389,6 +581,51 @@ def query_accessible_rows(self, cls, user_or_token, columns=None):
...
+            The mapped class of the target table.
+        user_or_token: `baselayer.app.models.User` or `baselayer.app.models.Token`
```
Comment (stefanv) [inline app/models.py:592]:
```suggestion
        user_or_token : `baselayer.app.models.User` or `baselayer.app.models.Token`
```

**Review by stefanv — COMMENTED** [review]
OK, great, I think this is just about there.  Let's just clarify Session vs VerifiedSession then we can get this in. Thanks!

File: app/models.py (hunk 1 above)
Comment (guynir42) [reply app/models.py:45]: It is renamed in the function call below the class definition.

File: app/models.py (hunk 4 above)
Comment (guynir42) [reply app/models.py:121]: Good call.

**Pull request merged**